File: android/app/src/main/kotlin/com/groovelab/groovelab/audio/RecordingEngine.kt
package com.groovelab.groovelab.audio

import android.Manifest
import android.content.Context
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import java.io.File
import java.io.FileOutputStream
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Recording engine with real-time onset detection for Android.
 * Uses AudioRecord for low-latency microphone capture.
 * Energy-based onset detection runs in the recording thread.
 */
class RecordingEngine(
    private val context: Context,
    private val onEvent: (Map<String, Any>) -> Unit
) {
    private val sampleRate = 44100
    private var audioRecord: AudioRecord? = null
    private var recordingThread: Thread? = null
    @Volatile private var isRecording = false
    private var currentFilePath: String? = null

    // Onset detection
    @Volatile private var onsetDetectionEnabled = false
    @Volatile private var onsetThreshold = 0.1
    @Volatile private var minOnsetIntervalMs = 50
    private var lastOnsetTimeNs: Long = 0
    private var previousEnergy: Double = 0.0
    private val energyHistory = mutableListOf<Double>()
    private val energyHistorySize = 10

    fun startRecording() {
        if (isRecording) return

        val bufferSize = AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        ) * 2

        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            bufferSize
        )

        // Create output file
        val dir = context.getExternalFilesDir(null) ?: context.filesDir
        val timestamp = System.currentTimeMillis()
        val file = File(dir, "take_$timestamp.wav")
        currentFilePath = file.absolutePath

        isRecording = true
        previousEnergy = 0.0
        energyHistory.clear()
        lastOnsetTimeNs = 0

        audioRecord?.startRecording()

        recordingThread = Thread({
            android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_URGENT_AUDIO)

            val buffer = ShortArray(512)
            val rawFile = File(dir, "take_${timestamp}_raw.pcm")
            val outputStream = FileOutputStream(rawFile)

            onEvent(mapOf(
                "type" to "recordingState",
                "isRecording" to true,
                "durationMs" to 0
            ))

            while (isRecording) {
                val read = audioRecord?.read(buffer, 0, buffer.size) ?: 0
                if (read > 0) {
                    // Write raw PCM
                    val byteBuffer = ByteBuffer.allocate(read * 2).order(ByteOrder.LITTLE_ENDIAN)
                    for (i in 0 until read) {
                        byteBuffer.putShort(buffer[i])
                    }
                    outputStream.write(byteBuffer.array())

                    // Onset detection
                    if (onsetDetectionEnabled) {
                        detectOnset(buffer, read)
                    }
                }
            }

            outputStream.close()

            // Convert raw PCM to WAV
            convertToWav(rawFile, File(currentFilePath!!), sampleRate)
            rawFile.delete()

        }, "GrooveLab-Recording").apply {
            priority = Thread.MAX_PRIORITY - 1
        }
        recordingThread?.start()
    }

    fun stopRecording(): String? {
        if (!isRecording) return null
        isRecording = false
        recordingThread?.join(1000)
        audioRecord?.stop()
        audioRecord?.release()
        audioRecord = null

        onEvent(mapOf(
            "type" to "recordingState",
            "isRecording" to false,
            "durationMs" to 0,
            "filePath" to (currentFilePath ?: "")
        ))

        return currentFilePath
    }

    private fun detectOnset(buffer: ShortArray, length: Int) {
        // Calculate RMS energy
        var sum = 0.0
        for (i in 0 until length) {
            val sample = buffer[i].toDouble() / 32768.0
            sum += sample * sample
        }
        val rms = sqrt(sum / length)

        // Spectral flux (energy increase)
        val flux = max(0.0, rms - previousEnergy)
        previousEnergy = rms

        // Running average for adaptive threshold
        energyHistory.add(flux)
        if (energyHistory.size > energyHistorySize) {
            energyHistory.removeAt(0)
        }
        val avgFlux = energyHistory.sum() / energyHistory.size
        val adaptiveThreshold = max(onsetThreshold, avgFlux * 2.5)

        if (flux > adaptiveThreshold) {
            val nowNs = System.nanoTime()
            val elapsedMs = (nowNs - lastOnsetTimeNs) / 1_000_000.0

            if (lastOnsetTimeNs == 0L || elapsedMs >= minOnsetIntervalMs) {
                lastOnsetTimeNs = nowNs

                // Peak amplitude
                var peak = 0.0
                for (i in 0 until length) {
                    peak = max(peak, abs(buffer[i].toDouble()) / 32768.0)
                }

                onEvent(mapOf(
                    "type" to "onset",
                    "timestampUs" to (nowNs / 1000),
                    "amplitude" to peak.coerceAtMost(1.0)
                ))
            }
        }
    }

    fun enableOnsetDetection(threshold: Double, minIntervalMs: Int) {
        onsetThreshold = threshold
        minOnsetIntervalMs = minIntervalMs
        onsetDetectionEnabled = true
    }

    fun disableOnsetDetection() {
        onsetDetectionEnabled = false
    }

    fun getInputLatency(): Double {
        val minBuffer = AudioRecord.getMinBufferSize(
            sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT
        )
        return (minBuffer.toDouble() / sampleRate) * 1000.0
    }

    companion object {
        fun convertToWav(pcmFile: File, wavFile: File, sampleRate: Int) {
            val pcmSize = pcmFile.length().toInt()
            val wavSize = pcmSize + 44

            val header = ByteBuffer.allocate(44).order(ByteOrder.LITTLE_ENDIAN)
            // RIFF
            header.put("RIFF".toByteArray())
            header.putInt(wavSize - 8)
            header.put("WAVE".toByteArray())
            // fmt
            header.put("fmt ".toByteArray())
            header.putInt(16) // chunk size
            header.putShort(1) // PCM
            header.putShort(1) // mono
            header.putInt(sampleRate)
            header.putInt(sampleRate * 2) // byte rate
            header.putShort(2) // block align
            header.putShort(16) // bits per sample
            // data
            header.put("data".toByteArray())
            header.putInt(pcmSize)

            val fos = FileOutputStream(wavFile)
            fos.write(header.array())
            fos.write(pcmFile.readBytes())
            fos.close()
        }
    }
}

File: groovelab_backup_20260325/ios/Runner/AudioEngine/RecordingEngine.swift
import AVFoundation
import Accelerate

/// Recording engine with real-time onset detection.
/// Uses AVAudioEngine input tap for low-latency mic capture
/// and energy-based onset detection for timing analysis.
final class RecordingEngine {
    private var audioEngine: AVAudioEngine?
    private var inputNode: AVAudioInputNode?
    private var audioFile: AVAudioFile?
    private var isRecording = false

    // Onset detection
    private var onsetDetectionEnabled = false
    private var onsetThreshold: Double = 0.1
    private var minOnsetIntervalMs: Int = 50
    private var lastOnsetTime: UInt64 = 0
    private var previousEnergy: Float = 0
    private var energyHistory: [Float] = []
    private let energyHistorySize = 10

    // Callback to send events to Flutter
    private let onEvent: ([String: Any]) -> Void

    // Recording file path
    private var currentFilePath: String?

    init(onEvent: @escaping ([String: Any]) -> Void) {
        self.onEvent = onEvent
    }

    // MARK: - Recording

    func startRecording() {
        guard !isRecording else { return }

        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.playAndRecord, options: [.defaultToSpeaker, .allowBluetooth])
            try session.setActive(true)
        } catch {
            print("RecordingEngine: AudioSession error: \(error)")
            return
        }

        let engine = AVAudioEngine()
        let input = engine.inputNode
        let format = input.outputFormat(forBus: 0)

        // Create output file
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let timestamp = Int(Date().timeIntervalSince1970 * 1000)
        let filePath = documentsPath.appendingPathComponent("take_\(timestamp).wav")
        currentFilePath = filePath.path

        do {
            audioFile = try AVAudioFile(forWriting: filePath, settings: [
                AVFormatIDKey: kAudioFormatLinearPCM,
                AVSampleRateKey: format.sampleRate,
                AVNumberOfChannelsKey: 1,
                AVLinearPCMBitDepthKey: 16,
                AVLinearPCMIsFloatKey: false,
            ])
        } catch {
            print("RecordingEngine: File creation error: \(error)")
            return
        }

        // Install tap on input for recording + onset detection
        input.installTap(onBus: 0, bufferSize: 512, format: format) { [weak self] buffer, time in
            self?.processAudioBuffer(buffer, time: time)
        }

        do {
            try engine.start()
        } catch {
            print("RecordingEngine: Engine start error: \(error)")
            return
        }

        self.audioEngine = engine
        self.inputNode = input
        isRecording = true
        previousEnergy = 0
        energyHistory.removeAll()
        lastOnsetTime = 0

        onEvent([
            "type": "recordingState",
            "isRecording": true,
            "durationMs": 0,
        ])
    }

    func stopRecording() -> String? {
        guard isRecording else { return nil }

        inputNode?.removeTap(onBus: 0)
        audioEngine?.stop()
        audioEngine = nil
        inputNode = nil
        audioFile = nil
        isRecording = false

        onEvent([
            "type": "recordingState",
            "isRecording": false,
            "durationMs": 0,
            "filePath": currentFilePath ?? "",
        ])

        return currentFilePath
    }

    // MARK: - Audio Processing

    private func processAudioBuffer(_ buffer: AVAudioPCMBuffer, time: AVAudioTime) {
        // Write to file
        if let file = audioFile {
            do {
                try file.write(from: buffer)
            } catch {
                print("RecordingEngine: Write error: \(error)")
            }
        }

        // Onset detection
        if onsetDetectionEnabled {
            detectOnset(buffer: buffer, time: time)
        }
    }

    // MARK: - Onset Detection

    /// Energy-based onset detection with spectral flux.
    /// Detects sudden increases in audio energy (transients)
    /// that correspond to note attacks.
    private func detectOnset(buffer: AVAudioPCMBuffer, time: AVAudioTime) {
        guard let channelData = buffer.floatChannelData else { return }
        let frameCount = Int(buffer.frameLength)
        let data = channelData[0]

        // Calculate RMS energy of this buffer
        var rms: Float = 0
        vDSP_rmsqv(data, 1, &rms, vDSP_Length(frameCount))

        // Calculate spectral flux (energy increase)
        let flux = max(0, rms - previousEnergy)
        previousEnergy = rms

        // Maintain running average for adaptive threshold
        energyHistory.append(flux)
        if energyHistory.count > energyHistorySize {
            energyHistory.removeFirst()
        }

        let avgFlux = energyHistory.reduce(0, +) / Float(energyHistory.count)
        let adaptiveThreshold = max(Float(onsetThreshold), avgFlux * 2.5)

        // Check if this is an onset
        if flux > adaptiveThreshold {
            var timebaseInfo = mach_timebase_info_data_t()
            mach_timebase_info(&timebaseInfo)
            let nsPerTick = Double(timebaseInfo.numer) / Double(timebaseInfo.denom)

            let nowHost = mach_absolute_time()
            let nowNs = Double(nowHost) * nsPerTick
            let lastNs = Double(lastOnsetTime) * nsPerTick
            let elapsedMs = (nowNs - lastNs) / 1_000_000

            // Enforce minimum interval between onsets
            if lastOnsetTime == 0 || elapsedMs >= Double(minOnsetIntervalMs) {
                lastOnsetTime = nowHost
                let timestampUs = Int(nowNs / 1000)

                // Calculate amplitude (peak in this buffer)
                var peak: Float = 0
                vDSP_maxv(data, 1, &peak, vDSP_Length(frameCount))
                let amplitude = min(1.0, Double(peak))

                onEvent([
                    "type": "onset",
                    "timestampUs": timestampUs,
                    "amplitude": amplitude,
                ])
            }
        }
    }

    func enableOnsetDetection(threshold: Double, minIntervalMs: Int) {
        onsetThreshold = threshold
        minOnsetIntervalMs = minIntervalMs
        onsetDetectionEnabled = true
    }

    func disableOnsetDetection() {
        onsetDetectionEnabled = false
    }

    func getInputLatency() -> Double {
        return AVAudioSession.sharedInstance().inputLatency * 1000.0 // ms
    }
}
